lib/layer.js

```diff
@@ -65,9 +78,23 @@ Layer.prototype.handle_error = function handle_error(error, req, res, next) {
     return next(error)
   }
 
+  req.layerStack = req.layerStack || []
+  req.layerStack.push(this)
+
+  if (onHandleError && onHandleError.shouldPublish()) {
+    onHandleError.publish({
+      error: error,
+      request: req,
+      response: res,
+      layer: this
+    })
+  }
+
   try {
     fn(error, req, res, next)
+    req.layerStack.pop()
   } catch (err) {
+    req.layerStack.pop()
```

Comment:
Can this just be put into a `finally` block?

Reply:
That'd run _after_ the `next(...)` though, which would interfere with subsequent routes, as far as I can tell. Could do that if the `next(...)` on the following line was in a `process.nextTick(...)`, but that'd have a performance hit.

Reply:
Oh. But isn't the one within the `try` already running after the `next` call? Like if the middleware sync calls `next` then it will run after, but if the middleware async calls `next` it will be before.

Reply:
Yeah, true. Probably a better place to put that code. 🤔

Reply:
Yea, it looks like adding anything non-trivial to your tests seems to highlight the weirdness here. For example, I added the `cors` module to the `nested routes` test like the following:
```js
  it('nested routes have multiple layers with paths', function (done) {
    var outer = new Router();
    var inner = new Router();

    inner.get('/:name', function (req, res) {
      res.end();
    });

    outer.use('/hello', cors(), inner);

    function end() {
      assert.strictEqual(joinLayerStack(handleRequest.request.layerStack), '/hello/:name/');
      done();
    }

    outer.handle({ url: '/hello/world', method: 'GET', headers: {} }, { end, setHeader: noop }, noop);
  });
```
But it failed the test listing the path as `/hello/hello/:name/`, as I would surmise that it is performing the pop operation too late, after the middleware is getting iterated through.

Reply:
I pushed a possible fix for the layer timing issue. Not sure it's the best solution, but seems to work. Still unsure if the layerStack approach is the best way. Willing to accept any suggestions on better ways to track the overall routing state of the app. 😅

Reply:
Ah, yea. It seems seems to have issues with the latest change keeping track of the stack. For example adding `inner.get('/world', (req, res, next) => next())` above `inner.get` produces the route as `/hello/world/:name`.

As far as a best solution... I think perhaps it may be better to maybe start from the top so we can understand what, exactly, we are trying to accomplish in this pull request? The PR title and description only mentions adding hooks for `diagnostics_channel`, but almost all the following discussion does not seem related to the actual `diagnostics_channel` usage, and rather adding a completely new feature, independent of the `diagnostics_channel` feature.

Should these two things be split apart into two PRs (first suggested at https://github.com/pillarjs/router/pull/96#issuecomment-694917009) or can the description be updated to describe what the goal of this PR is? That may help better focus the conversation and code work.

Reply:
Capturing routing information is the whole _point_ of adding `diagnostics_channel`, so this PR doesn't really make sense without it. I'll update the description to elaborate on that need.

Reply:
Yea, I get it, but let's think iteratively here, if you are interested in getting things landed instead of waiting for the entire pie to be built. For example, folks want to construct the path even outside of `diagnostics_channel`, so having access to the path is an independent ask of adding `diagnostics_channel` support. In addition, if you didn't have that layer stack hanging off the `req` object, you still already have access to a whole bunch of information, including that the request is being handled, that an error occurred, what the http method is, the http path, etc.

Reply:
We get all that directly from http already. The only context we're really missing from express is the routing information. I only intended the layer tracking stuff in this to be used by diagnostics_channel. Maybe there are more general-purpose uses for that information, but I lack the context and the time to do anything about those right now. This PR was created primarily as a demonstration that the `diagnostics_channel` API is usable in more complex scenarios. It's not intended to be landed any time soon as the `diagnostics_channel` feature itself hasn't even landed yet. If you want something more complex from this, I'm probably not going to get to it for awhile.